src/text/non_blocking/raw_text_reader.rs

```diff
@@ -1261,4 +1408,91 @@ mod reader_tests {
         assert!(result.is_err());
         Ok(())
     }
+
+    // Test stepping out at the top level with incomplete error during step_out call.
+    #[test]
+    fn resume_step_out() -> IonResult<()> {
+        // The spacing on this is important, since we need to know where (approximately) the
+        // incomplete error occurs.
+        let source = r#"{field:{another_field:{foo:"We should fail somewhere in this string.."}}}"#;
+
+        let mut reader = RawTextReader::new(source[..47].as_bytes().to_owned());
+        let result = reader.next()?;
+        assert_eq!(result, RawStreamItem::Value(IonType::Struct));
+
+        reader.step_in()?;
+        assert_eq!(reader.depth(), 1);
+
+        reader.next()?;
+
+        reader.step_in()?;
+        assert_eq!(reader.depth(), 2);
+
+        match reader.step_out() {
+            Err(IonError::IncompleteText { .. }) => {
+                assert_eq!(reader.depth(), 3); // we should fail at depth 3, within the foo field.
+                reader.read_from(&mut source[47..].as_bytes(), 512)?;
+
+                // Resume our initial step_out, this should bring us back to depth 1.
+                reader.step_out()?;
+                assert_eq!(reader.depth(), 1);
```

Comment:
If I'm at depth 3 and `step_out` fails, then the next time I call `step_out` successfully I should be at depth 2. If I'm understanding this correctly, the current implementation allows the reader to make _some_ progress before failing and the user can then resume using it from that unknown interstitial position. I think there's a lot of room for confusion here (how far did the reader get? where is it now?), and it's not consistent with our other `Reader` implementations' contracts.

Reply:
Sorry for the sporadic responses, I let this one marinate for a bit. Your comment about the reader implementations' contracts is very surprising.

Confusion for whom? The user? Or maintainers?
What reader contract? I may have missed documentation describing this.

If the intent truly results in requiring the entire remaining content of a container to be in memory before a `step_out` can actually make progress, and re-parsing the same data until it does, then I have concerns. Let's discuss.

Reply:
Update after our discussion:

I misunderstood Zach's comment. We do not need to roll back to where the reader was when `step_out`/`next` was called, we need to ensure that if we cannot complete the `step_out` or `next` then we do not allow the user to advance the reader in any other way until the initial failed call is completed. This should match the behavior of the binary reader.